routing_common/transit_serdes.hpp

```diff
@@ -0,0 +1,128 @@
+#pragma once
+
+#include "geometry/point2d.hpp"
+
+#include "coding/point_to_integer.hpp"
+#include "coding/read_write_utils.hpp"
+#include "coding/reader.hpp"
+#include "coding/varint.hpp"
+#include "coding/write_to_sink.hpp"
+
+#include "base/assert.hpp"
+#include "base/exception.hpp"
+#include "base/macros.hpp"
+
+#include <cmath>
+#include <cstdint>
+#include <limits>
+#include <sstream>
+#include <string>
+#include <type_traits>
+#include <vector>
+
+namespace routing
+{
+namespace transit
+{
+template <typename Sink>
+class Serializer
+{
+public:
+  Serializer(Sink & sink) : m_sink(sink) {}
+
+  template<typename T>
+  typename std::enable_if<std::is_integral<T>::value || std::is_enum<T>::value>::type
+      operator()(T const & t, char const * /* name */ = nullptr)
+  {
+    WriteToSink(m_sink, t);
+  }
+
+  template<typename T>
+  typename std::enable_if<std::is_floating_point<T>::value>::type operator()(T const & t, char const * /* name */ = nullptr)
+  {
+    NOTIMPLEMENTED();
+  }
+
+  void operator()(std::string const & s, char const * /* name */ = nullptr)
+  {
+    rw::Write(m_sink, s);
+  }
+
+  void operator()(m2::PointD const & p, char const * /* name */ = nullptr)
+  {
+    m2::PointU const pointU = PointD2PointU(p, POINT_COORD_BITS);
+    (*this)(pointU.x);
```

Comment:
In general, it's a good idea to write pointU.x and pointU.y as varints, relative to mercator base.

Reply:
Or, relative to a previous point.

Reply:
Are you talking about something like that:
```
    WriteVarUint(m_sink, pointU.x);
    WriteVarUint(m_sink, pointU.y);
```
and
```
    pointU.x = ReadVarUint<uint32_t, Source>(m_source);
    pointU.y = ReadVarUint<uint32_t, Source>(m_source);
```
?
Done

Reply:
Yes, you can reuse EncodeDelta from the last encoded point.

Reply:
EncodeDelta()? But I don't have any |prediction| to use it. 

But I can save a point of mwm center in header. In that case I do have a |prediction|. Do you propose to do that? 

To implement this I need to keep mwm center point at serializer/deserializer.

Reply:
You can use last encoded point as a prediction for a new one. This is a poor prediction, but better than nothing.

Reply:
I propose leave it as it for now. The problem is I'm planning to use Serializer/Deserializer for reading all table at transit section. Router will read some set of tables of transit section. And drape will read another set (or ever two sets). This 2-3 readers may be called in different order. And the value to last encoded point depend on the order. So I cannot cache easily last encoded point at Serializer/Deserializer to use it in prediction.